### PR TITLE
Implemented ChipAtPos tactic

### LINKED ISSUE #858: ChipAtPos
Same as #855 except with a chip

### PR DESCRIPTION
**Pull request by timovdk**
Implemented the ChipAtPos tactic #858

### CODE REVIEW

File: src/stp/new_skills/Rotate.cpp [hunk 1]
```diff
@@ -19,7 +19,7 @@ Status Rotate::onUpdate(const StpInfo &info) noexcept {
     int targetDribblerSpeed = targetDribblerPercentage / 100.0 * Constants::MAX_DRIBBLER_CMD();
 
     // Set angle command
-    command.set_w(targetAngle.getAngle());
+    command.set_w(targetAngle);
```
Comment (jordiluong) [inline src/stp/new_skills/Rotate.cpp]:
```suggestion
    command.set_w(targetAngle.getAngle());
```
Angle class

File: src/stp/new_skills/Rotate.cpp [hunk 2]
```diff
@@ -28,7 +28,7 @@ Status Rotate::onUpdate(const StpInfo &info) noexcept {
 
     // Check if successful
     double errorMargin = Constants::GOTOPOS_ANGLE_ERROR_MARGIN() * M_PI;
-    if (info.getRobot().value()->getAngle().shortestAngleDiff(targetAngle) < errorMargin) {
+    if (fabs(info.getRobot().value()->getAngle() - targetAngle) < errorMargin) {
```
Comment (jordiluong) [inline src/stp/new_skills/Rotate.cpp]:
```suggestion
    if (info.getRobot().value()->getAngle().shortestAngleDiff(targetAngle) < errorMargin) {
```
Angle class

Comment (timovdk) [inline src/stp/new_skills/Rotate.cpp]:
```suggestion
    if (fabs(info.getRobot().value()->getAngle().shortestAngleDiff(targetAngle)) < errorMargin) {
```

File: include/roboteam_ai/stp/StpInfo.h [hunk 3]
```diff
@@ -105,6 +117,9 @@ struct StpInfo {
     int getDribblerSpeed() const { return dribblerSpeed; }
     void setDribblerSpeed(int dribblerSpeed) { this->dribblerSpeed = dribblerSpeed; }
 
+    KickChipType getKickChipType() const { return kickChipType; }
+    void setKickChipType(KickChipType kickChipType) { StpInfo::kickChipType = kickChipType; }
```
Comment (Ratoone) [inline include/roboteam_ai/stp/StpInfo.h]: Not that familiar with cpp, but doesn't this imply that the field is static?

File: src/stp/Play.cpp [hunk 4]
```diff
@@ -25,17 +25,20 @@ Status Play::update() noexcept {
...
 
+    calculateInfoForPlay();
+
     for (auto& each : roles) {
         auto roleName{each->getName()};
         if(stpInfos.find(roleName) != stpInfos.end()) {
-            // TODO refresh robots
+            // TODO refresh views
             stpInfos.find(roleName)->second.setRobot(world->getWorld()->getRobotForId(stpInfos.find(roleName)->second.getRobot()->get()->getId()));
+            stpInfos.find(roleName)->second.setBall(world->getWorld()->getBall());
```
Comment (Ratoone) [inline src/stp/Play.cpp]: This was already implemented development

File: src/stp/new_plays/TestPlay.cpp [hunk 5]
```diff
@@ -49,22 +49,19 @@ void TestPlay::assignRoles() noexcept {
             stpInfos.emplace(roleName, StpInfo{});
             stpInfos[roleName].setRobot(robot);
             stpInfos[roleName].setField(*world->getField());
-
-            // TODO calculate additional info
+            stpInfos[roleName].setBall(*world->getWorld()->getBall());
```
Comment (Ratoone) [inline src/stp/new_plays/TestPlay.cpp]: These things already happen in the base play

File: src/stp/new_skills/Chip.cpp [hunk 6]
```diff
@@ -21,7 +21,12 @@ Status Chip::onUpdate(const StpInfo &info) noexcept {
 
     publishRobotCommand();
 
-    return Status::Success;
+    if(info.getBall()->get()->getVelocity().length() > Constants::BALL_STILL_VEL()) {
+        return Status::Success;
+    }
+    else {
+        return Status::Running;
+    }
```
Comment (Ratoone) [inline src/stp/new_skills/Chip.cpp]: Redundant else

```suggestion
    return Status::Running;
```

File: src/stp/new_tactics/ChipAtPos.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    for (auto &x : skills) {
+        x->terminate();
+    }
+}
+
+StpInfo ChipAtPos::calculateInfoForSkill(StpInfo const &info) noexcept {
+    if (!info.getBall() || !info.getRobot() || !info.getField()) {
+        RTT_WARNING("No Ball, Robot or Field present in StpInfo");
+        return {};
+    }
```
Comment (Ratoone) [inline src/stp/new_tactics/ChipAtPos.cpp]: Present in base class already. Check development

File: src/stp/new_tactics/ChipAtPos.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    // Calculate the angle the robot needs to aim
+    double angleToBall = (info.getPosition().second - info.getBall()->get()->getPos()).angle();
+    skillStpInfo.setAngle(angleToBall);
+
+    // Calculate the distance and the chip force
+    double distanceBallToTarget = (info.getBall()->get()->getPos() - info.getPosition().second).length();
+    skillStpInfo.setKickChipVelocity(determineChipForce(distanceBallToTarget, skillStpInfo.getKickChipType()));
+
+    // When chipping, we don't need to activate the dribbler
+    skillStpInfo.setDribblerSpeed(0);
```
Comment (Ratoone) [inline src/stp/new_tactics/ChipAtPos.cpp]: We do while rotating with the ball

File: src/stp/new_tactics/ChipAtPos.cpp [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            break;
+        }
+        default: {
+            velocity = rtt::ai::Constants::MAX_KICK_POWER();
+        }
+    }
+
+    // limit the output to the max chip speed
+    return std::min(std::max(velocity, 1.01), rtt::ai::Constants::MAX_KICK_POWER());
+}
```
Comment (Ratoone) [inline src/stp/new_tactics/ChipAtPos.cpp:90]: Warning: magic numbers detected. Magic is forbidden in our code. Also, can you write some documentation of comments why grsim is special? As I can see, you just divide the speed by 2, so maybe it's better to do that in the cases.

**Review by RobotJente — APPROVED** [review]
Looks good

**Review by Ratoone — CHANGES_REQUESTED** [review]
Using magic numbers is crime against the rest of the team, punishable by death. Please do something about that and document the complex computations

File: src/stp/new_tactics/ChipAtPos.cpp (hunk 7 above)
Comment (timovdk) [reply src/stp/new_tactics/ChipAtPos.cpp]: woops forgot to remove this, added this before the meeting this morning

File: src/stp/Play.cpp (hunk 4 above)
Comment (timovdk) [reply src/stp/Play.cpp]: i know

File: include/roboteam_ai/stp/StpInfo.h (hunk 3 above)
Comment (timovdk) [reply include/roboteam_ai/stp/StpInfo.h]: o woops, autmatically generated setters aren't great

File: src/stp/new_tactics/ChipAtPos.cpp (hunk 9 above)
Comment (timovdk) [reply src/stp/new_tactics/ChipAtPos.cpp:90]: We definitely should refactor this for both kicking and chipping

File: src/stp/new_tactics/ChipAtPos.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    double angleToBall = (info.getPosition().second - info.getBall()->get()->getPos()).angle();
+    skillStpInfo.setAngle(angleToBall);
+
+    // Calculate the distance and the chip force
+    double distanceBallToTarget = (info.getBall()->get()->getPos() - info.getPosition().second).length();
+    skillStpInfo.setKickChipVelocity(determineChipForce(distanceBallToTarget, skillStpInfo.getKickChipType()));
+
+    // When rotating, we need to dribble to keep the ball, but when kicking we don't
+    if(skills.current_num() == 0) {
+        skillStpInfo.setDribblerSpeed(31);
```
Comment (timovdk) [inline src/stp/new_tactics/ChipAtPos.cpp]:
```suggestion
        skillStpInfo.setDribblerSpeed(100);
```

File: src/stp/new_plays/TestPlay.cpp [hunk 11]
```diff
@@ -52,15 +52,16 @@ void TestPlay::assignRoles() noexcept {
...
         for (auto & role : roles) {
             auto roleName{role->getName()};
             if (stpInfos.find(roleName) != stpInfos.end()) {
                 auto robot = stpInfos[roleName].getRobot().value();
                 // TODO calculate additional info
+                stpInfos[roleName].setPosition({SHOOT_TO_POSITION, {robot->getId()*0.2, robot->getId()*0.2}});
                 stpInfos[roleName].setEnemyRobot(world->getWorld()->getThem()[0]);
                 stpInfos[roleName].setBlockDistance(BlockDistance::HALFWAY);
-                stpInfos[roleName].setPosition({SHOOT_TO_POSITION, {robot->getId()*0.2, robot->getId()*0.2}});
+
```
Comment (timovdk) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
```

**Pull request merged**